fix: hide send message/unfriend buttons on friend list + tests

darrenvong opened the pull request:
And some general clean ups from hardcoded messages to constants as agreed.

Closes #667

darrenvong committed:
fix: hide send message/unfriend buttons on friend list + tests

darrenvong committed:
fix: clear friend requests queries on logout

darrenvong committed:
fix: disable react query cache

Seems to be a weird bug with the experimental plugin where user
sensitive data like messages and friends are not purged on logout

lucaslcode commented on app/frontend/src/features/auth/Logout.tsx:
```diff
@@ -17,6 +18,8 @@ export default function Logout() {
       queryClient.resetQueries("hostRequests");
       queryClient.resetQueries("ping");
+      queryClient.resetQueries(friendRequestKey("Sent"));
+      queryClient.resetQueries(friendRequestKey("Received"));
```
I think we should just clear absolutely everything. That might fix the offline cache bug too.

I mean, even the user cache shouldn't be left on a public computer (for example) after logging out, if users are only visible to couchers members

darrenvong replied:
Yeah that might make this more maintainable going forward tbh, otherwise it's easy to forget to remove a query here. I guess there's probably still value in having the cache while the user is logged in so that between page refreshes it'll still be faster than no cache

lucaslcode commented on app/frontend/src/queryKeys.ts:
```diff
@@ -40,3 +40,7 @@ export const referencesKey = (
+
+export type FriendRequestType = "Sent" | "Received";
+export const friendRequestKey = (type: FriendRequestType) =>
+  `friendRequests${type}`;
```
```suggestion
  ["friendRequests", type];
```
Second param could even be an object, since it's "filters".

lucaslcode commented on app/frontend/src/reactQueryClient.tsx:
```diff
@@ -22,13 +20,6 @@ export const queryClient = new QueryClient({
   },
 });
 
-const persistor = createLocalStoragePersistor();
-
-persistQueryClient({
-  maxAge: 14 * 24 * 60 * 60 * 1000,
-  persistor,
-  queryClient,
-});
```
Did you mean to commit this?

lucaslcode replied:
Oh I see your commit message. I will check out the plugin source

lucaslcode replied:
https://github.com/tannerlinsley/react-query/issues/1951

lucaslcode commented on app/frontend/src/test/serviceMockDefaults.ts:
```diff
@@ -1,11 +1,12 @@
 
 export async function getUser(userId: string) {
-  return userMap.get(+userId);
+  return userMap.get(+userId) as User.AsObject;
```
Why this change?

darrenvong replied:
It's because I've used the real return type of `getUser` and the fixtures don't actually have the full user objects (a lot of missing keys deemed required in the type). Maybe I'll just add in everything as in the default user and stop being lazy 😅  I imagine eventually once we have better test coverage, we'll need those properties on the user anyway. Since it's already in a separate JSON file, I don't see it being a problem (mainly bloating up the test file with properties you don't care about in a specific test suite if it were inlined)

darrenvong replied:
Oh crap actually, might still need this or a non-null assertion since `userMap.get` returns `User.AsObject | undefined` but it should never be undefined in our case as we should always be using known userIds in our tests.

lucaslcode reviewed (changes requested):
Can you remove the all tab while you're in there?

lucaslcode commented:
[Looking here](https://github.com/tannerlinsley/react-query/blob/master/src/createLocalStoragePersistor-experimental/index.ts), it seems the cache updating is throttle to only happen every seconds, so I guess it's a race condition?

As a workaround, maybe set the local storage key to a constant, and then directly remove that from local storage on logout?

https://github.com/tannerlinsley/react-query/issues/1951

darrenvong committed:
fix: address all PR points

- friendRequests key to use ['friendRequests', { type }] instead of
simple string
- update user fixtures to have complete user data
- Removed "all" tab under connections

darrenvong commented:
The workaround of interacting with directly in logout didn't work :( I suspect this might also be a bottleneck on how quickly local storage can accept write to disk? 🤷

lucaslcode commented on app/frontend/src/reactQueryClient.tsx:
```diff
@@ -22,13 +20,6 @@ export const queryClient = new QueryClient({
   },
 });
 
-const persistor = createLocalStoragePersistor();
```
Adding `throttleTime: 100` fixed the issue for me.

lucaslcode reviewed (changes requested):
Really don't want to get rid of offline cache sorry :P But this worked for me

darrenvong committed:
fix: lower throttle time on offline cache persistor

darrenvong commented:
> Really don't want to get rid of offline cache sorry :P But this worked for me

Lol didn't know why I didn't try that first. Glad it was only a race condition and not some kinda disk write bottleneck! Fixed!
